File: unidad4.4_GRASP/calculator.cpp
#include <iostream>
using namespace std;

// Modelo
class Calculator {
public:
    float add(float a, float b) { return a + b; }
    float subtract(float a, float b) { return a - b; }
    float multiply(float a, float b) { return a * b; }
    float divide(float a, float b) {
        if (b == 0) {
            cout << "Error: division por cero." << endl;
            return 0;
        }
        return a / b;
    }
};

// controlador, coordina entre la vista y el modelo
class CalculatorController {
private:
    Calculator model;
public:
    float executeOperation(int option, float a, float b) {
        switch (option) {
            case 1: return model.add(a, b);
            case 2: return model.subtract(a, b);
            case 3: return model.multiply(a, b);
            case 4: return model.divide(a, b);
            default:
                cout << "Opción invalida." << endl;
                return 0;
        }
    }
};

// vista, interfaz de usuario
int main() {
    float number1, number2;
    int option;

    cout << "Ingrese el primer numero: ";
    cin >> number1;
    cout << "Ingrese el segundo numero: ";
    cin >> number2;
    cout << "Seleccione operacion: 1-Suma 2-Resta 3-Multiplicacion 4-Division: ";
    cin >> option;

    CalculatorController controller;
    float result = controller.executeOperation(option, number1, number2);
    cout << "Resultado: " << result << endl;

    return 0;
}

File: unidad4.4_GRASP/calculator.py
# Modelo
class Calculator:
    def add(self, a, b):
        return a + b

    def subtract(self, a, b):
        return a - b

    def multiply(self, a, b):
        return a * b

    def divide(self, a, b):
        if b == 0:
            print("Error: division por cero.")
            return 0
        return a / b

# Controlador
class CalculatorController:
    def __init__(self):
        self.model = Calculator()

    def execute_operation(self, option, a, b):
        if option == 1:
            return self.model.add(a, b)
        elif option == 2:
            return self.model.subtract(a, b)
        elif option == 3:
            return self.model.multiply(a, b)
        elif option == 4:
            return self.model.divide(a, b)
        else:
            print("Opcion invalida.")
            return 0

# Vista, interfaz de usuario
def main():
    number1 = float(input("Ingrese el primer numero: "))
    number2 = float(input("Ingrese el segundo numero: "))
    option = int(input("Seleccione operacion: 1-Suma 2-Resta 3-Multiplicacion 4-Division: "))

    controller = CalculatorController()
    result = controller.execute_operation(option, number1, number2)
    print("Resultado:", result)

main()